translib/version.go

```diff
@@ -16,24 +40,144 @@ var theYangBaseVersion Version
+	if err == nil && v.IsNull() {
+		err = fmt.Errorf("Invalid version \"%s\"", s)
+	}
+
+	return err
+}
+
+// IsNull checks if the version v is a null version (0.0.0)
+func (v *Version) IsNull() bool {
+	return v == nil || (v.Major == 0 && v.Minor == 0 && v.Patch == 0)
+}
+
+// GreaterThan checks if the Version v is more than another version
+func (v *Version) GreaterThan(other Version) bool {
+	return (v.Major > other.Major) ||
+		(v.Major == other.Major && v.Minor > other.Minor) ||
+		(v.Major == other.Major && v.Minor == other.Minor && v.Patch > other.Patch)
+}
+
+// GetCompatibleBaseVersion returns the compatible base version for
+// current version.
+func (v *Version) GetCompatibleBaseVersion() Version {
+	switch v.Major {
+	case 0:
+		// 0.x.x versions are always considered as developement versions
+		// and are not backward compatible. So, base = current
+		return *v
+	case 1:
+		// Base is 1.0.0 if yang bundle version is 1.x.x
+		return Version{Major: 1}
+	default:
+		// For 2.x.x or higher versions the base will be (N-1).0.0
+		return Version{Major: v.Major - 1}
+	}
+}
+
+// validateClientVersion verifies API client client is compatible with this server.
+func validateClientVersion(clientVer Version, path string, appInfo *appInfo) error {
+	if clientVer.IsNull() {
+		// Client did not povide version info
+		return nil
+	}
+
+	if appInfo.isNative {
+		return validateClientVersionForNonYangAPI(clientVer, path, appInfo)
+	}
+
+	return validateClientVersionForYangAPI(clientVer, path, appInfo)
+}
+
+// validateClientVersionForYangAPI checks theYangBaseVersion <= clientVer <= theYangBundleVersion
+func validateClientVersionForYangAPI(clientVer Version, path string, appInfo *appInfo) error {
+	if theYangBaseVersion.GreaterThan(clientVer) {
+		glog.Errorf("Client version %s is less than base version %s", clientVer, theYangBaseVersion)
+	} else if clientVer.GreaterThan(theYangBundleVersion) {
+		glog.Errorf("Client version %s is more than server vesion %s", clientVer, theYangBundleVersion)
+	} else {
+		return nil
+	}
+
+	return tlerr.TranslibUnsupportedClientVersion{
+		ClientVersion:     clientVer.String(),
+		ServerVersion:     theYangBundleVersion.String(),
+		ServerBaseVersion: theYangBaseVersion.String(),
+	}
+}
+
+// validateClientVersionForNonYangAPI checks client version for non-yang APIs
+func validateClientVersionForNonYangAPI(clientVer Version, path string, appInfo *appInfo) error {
+	// Version checks are not supported for non-yang APIs.
+	// Client versions are ignored.
+	return nil
+}
+
+//========= initialization =========
+
+func init() {
+	path := filepath.Join(GetYangPath(), "version.xml")
+	if err := initYangVersionConfigFromFile(path); err != nil {
+		glog.Errorf("Error loading version config file; err=%v", err)
+		glog.Errorf("API VERSION CHECK IS DISABLED")
+	} else {
+		glog.Infof("*** Yang bundle version = %s", theYangBundleVersion)
+		glog.Infof("*** Yang base version   = %s", theYangBaseVersion)
```

Comment:
The version XML has only bundle version. Where the base version come from ?

Curious! What is the difference between  bundle & base version?

Reply:
Renuka, it is documented in the code itself. We want to support one version below and hence that will be the base version that we will support.

 37 // theYangBaseVersion indicates the minimum yang bundle version
 38 // supported by translib APIs. This will be usually 1 major version
 39 // below current yang bundle version {theYangBundleVersion.Major-1, 0, 0}
 40 var theYangBaseVersion Version